ext/pdo/pdo_stmt.c

```diff
@@ -688,13 +634,41 @@ static bool pdo_do_key_pair_fetch(pdo_stmt_t *stmt, enum pdo_fetch_orientation o
 	return true;
 }
 
+/* Return value MUST be an initialized object */
+static bool pdo_call_fetch_object_constructor(zend_function *constructor, HashTable *ctor_args, zval *return_value)
+{
+	zval retval_constructor_call;
+	zend_fcall_info fci = {
+		.size = sizeof(zend_fcall_info),
+		.function_name = {},
+		.object = Z_OBJ_P(return_value),
+		.retval = &retval_constructor_call,
+		.param_count = 0,
+		.params = NULL,
+		.named_params = ctor_args,
+	};
+	zend_fcall_info_cache fcc = {
+		.function_handler = constructor,
+		.object = Z_OBJ_P(return_value),
+		.called_scope = Z_OBJCE_P(return_value),
+		.calling_scope = NULL,
+		.closure = NULL,
+	};
+
+	zend_call_function(&fci, &fcc);
+	bool failed = Z_ISUNDEF(retval_constructor_call);;
+	zval_ptr_dtor(&retval_constructor_call);
+
+	return failed;
+}
+
 /* perform a fetch.
  * Stores values into return_value according to HOW. */
 static bool do_fetch(pdo_stmt_t *stmt, zval *return_value, enum pdo_fetch_type how, enum pdo_fetch_orientation ori, zend_long offset, zval *group_key) /* {{{ */
```

Comment:
In `pdo_stmt_iter_move_forwards` you should set `I->fetch_ahead` to UNDEF too after its destruction to avoid a double destruction in case this function errors. Please document the invariant that `return_value` must be safely destructable (e.g. being UNDEF or NULL)